astropy/io/fits/hdu/image.py

```diff
@@ -264,19 +264,16 @@ def data(self, data):
             self._data_replaced = True
             was_unsigned = False
 
-        if (
-            data is not None
-            and not isinstance(data, np.ndarray)
-            and not _is_dask_array(data)
-        ):
-            # Try to coerce the data into a numpy array--this will work, on
-            # some level, for most objects
-            try:
-                data = np.array(data)
-            except Exception:
-                raise TypeError(
-                    f"data object {data!r} could not be coerced into an ndarray"
-                )
+        if data is not None:
```

Comment:
I don't see any real changes except you broke the if into two lines. Am I missing something real subtle here?

Reply:
Yes, it is a subtle change indeed. There's a check for `shape` here:

https://github.com/astropy/astropy/blob/99af4c9f51d417bcea28af310fe2cec158b1a0ea/astropy/io/fits/hdu/image.py#L278-L281

This check only occured when the three conditions in the if occurred (`data is not None and not isinstance(data, np.ndarray) and not _is_dask_array(data)`). With this changes, the `shape` check will always occur as long as `data is not None`.

Reply:
In other words, previously if we set `data` with a zero-dimensional array it would skip the shape test, but now it doesn't.

Reply:
Ah, I see it now. The shape check was outside the green diff. Thanks!